src/btree/bt_sync.c

```diff
@@ -153,6 +157,9 @@ __sync_file(WT_SESSION_IMPL *session, int syncop)
 				++leaf_pages;
 			}
 			WT_ERR(__wt_reconcile(session, walk, NULL, 0));
+
+			if (TXNID_LT(btree->rec_max_txn, mod->rec_max_txn))
+				btree->rec_max_txn = mod->rec_max_txn;
```

Comment:
That change won't deal with dirty pages that are written (and become clean) during the checkpoint.  I'll add comments regardless.